layers.go

```diff
@@ -602,6 +606,66 @@ func (r *layerStore) Status() ([][2]string, error) {
 	return r.driver.Status(), nil
 }
 
+func (r *layerStore) PutFromAdditionalLayerStore(id string, parentLayer *Layer, d digest.Digest, imageref string) (layer *Layer, err error) {
+	adriver, ok := r.driver.(drivers.AdditionalLayerStoreDriver)
+	if !ok {
+		return nil, ErrLayerUnknown
+	}
+
+	var diffDir string
+	var layerInfo *Layer
+	for _, ls := range adriver.AdditionalLayerStores() {
+		if layerInfo, diffDir, err = searchLayer(ls, d, imageref); err == nil {
+			break
+		} else if err != ErrLayerUnknown {
+			return nil, err
+		}
+	}
+
+	parent := ""
+	if parentLayer != nil {
+		parent = parentLayer.ID
+	}
+	if err = adriver.CreateFromDirectory(id, parent, diffDir); err != nil {
+		if id != "" {
+			return nil, errors.Wrapf(err, "error creating layer from dir with ID %q", id)
+		}
+		return nil, errors.Wrapf(err, "error creating layer from dir")
+	}
+
+	compressedsums := make(map[digest.Digest][]string)
+	uncompressedsums := make(map[digest.Digest][]string)
+	names := make(map[string]*Layer)
+	layer = copyLayer(layerInfo)
+	layer.ID = id
+	layer.Parent = parent
+	layer.Created = time.Now().UTC()
+	if layer.CompressedDigest != "" {
+		compressedsums[layer.CompressedDigest] = append(compressedsums[layer.CompressedDigest], layer.ID)
+	}
+	if layer.UncompressedDigest != "" {
+		uncompressedsums[layer.UncompressedDigest] = append(uncompressedsums[layer.UncompressedDigest], layer.ID)
+	}
+	for _, name := range layer.Names {
+		if conflict, ok := names[name]; ok {
+			r.removeName(conflict, name)
+		}
+		names[name] = layer
+	}
+	// TODO: check if necessary fields are filled
+	r.layers = append(r.layers, layer)
+	r.idindex.Add(id)
+	r.byid[id] = layer
+	r.byname = names
+	r.bycompressedsum = compressedsums
+	r.byuncompressedsum = uncompressedsums
```

Comment:
AFAICS this discards a lot of data about all the other layers. How does that work??!

Reply:
Fixed this.

Reply:
Looks better now (but needs a review by someone who knows c/storage, i.e. not me).